### ISSUE TITLE
OpenCV3 failed to build on MacOS 10.12

### ISSUE THREAD

**Issue by danieldf** [Homebrew/homebrew-core#5352]
- [x] If Homebrew was updated on Aug 10-11th 2016 and `brew update` always says `Already up-to-date.` you need to run: `cd "$(brew --repo)" && git fetch && git reset --hard origin/master && brew update`.
# Please follow the general troubleshooting steps first:
- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?
- [x] If you're seeing permission errors tried running `sudo chown -R $(whoami) $(brew --prefix)`?
### Bug reports:

Update, or new installation, of OpenCV 3 (3.1.0_4) fails at 3%; the full command line is: `brew install opencv3 --with-contrib --with-cuda --with-ffmpeg --with-tbb --with-qt5 --with-examples --with-java --with-opengl --with-quicktime`. All relevant downloads are successful (main tarballs and patches), and all patches are applied without errors. `cmake` runs, and `make` runs just after that: `make` exists with the following error code, `make: *** [all] Error 2`, right after `[  3%] Built target opencv_cvv_automoc`.

There were no issues installing, nor updating, OpenCV3 before the recent XCode update (using the very same `brew install` line as above).
- [`brew gist-logs opencv3`](https://gist.github.com/anonymous/1dc84bfa40611eb0c2138d28f8a5f6de).

**Comment by dunn** [comment]
Looks like they just need to update their scripts to allow Xcode 8:
```
nvcc fatal   : The version ('80000') of the host compiler ('Apple clang') is not supported
CMake Error at cuda_compile_generated_gpu_mat.cu.o.cmake:206 (message):
  Error generating
/tmp/opencv3-20160928-96456-17tvsa5/opencv-3.1.0/macbuild/modules/core/CMakeFiles/cuda_compile.dir/src/cuda/./cuda_compile_generated_gpu_mat.cu.o

**Comment by danieldf** [comment]
@dunn: is this something that i can do myself (in my own installation)? Or should i wait until the Homebrew team does the job?

**Comment by dunn** [comment]
It's something the OpenCV team needs to fix on their end.  If you let them know (or +1 an existing ticket), that might help speed things up!

**Comment by virajago** [comment]
@dunn I face the same error after make runs completing 9%
==> make
Generating moc source moc_zoomableimage.cpp
.......
Generating moc compilation opencv_cvv_automoc.cpp
[  9%] Built target opencv_cvv_automoc
make: *** [all] Error 2

**Comment by myfavouritekk** [comment]
I think it's CUDA 8's problem. Basically, CUDA 8.0 has not supported Xcode 8 yet and all packages relying on CUDA need to temporarily use Xcode 7.3 instead. However, Homebrew requires newest Xcode 8.0. It is stuck there util Nvidia updates CUDA 8 to support Xcode 8.

**Comment by virajago** [comment]
so, is there any version of opencv that can installed with Xcode 8 and CUDA 8? I have Xcode 7.3 as well which I use to run Cuda 8

**Comment by danieldf** [comment]
@virajago: i spent a significant amount of time on this issue yesterday... and there really doesn't seem to be a way out of it. Except for the following:

1. Update/Upgrade all of your HomeBrew installation, and also install all relevant dependencies to compile OpenCV by hand;
2. Uninstall your XCode 8 (even if you're on MacOS Sierra, 10.12): move it to the Trash folder, empty/delete it, etc;
3. Download and install XCode 7.3 (i think 7.3 was the previous version) and its associated Command Line Tools;
4. Download and compile OpenCV *by hand*, then install it appropriately (copy 'cv.py' and 'cv2.so' to the correct Python locations, and use 'make install' for everything else): compilation will take a long time (maybe ~1h, you can use 'make -j 3' or so to speed things up), but should work just fine;
5. Remove XCode 7.3, and **reinstall** XCode 8 and its associated Command Line Tools;
6. Update/Upgrade your 'brew' install as you normally would: everything HomeBrew-related should work correctly once XCode 8 is back in place.

So, once you compile and install OpenCV (by hand) using the previous version of XCode, it'll still work correctly after you upgrade XCode to version 8, and so on. If you want, before all this, you can also upgrade to the newer CUDA 8 (as longs as you compile with XCode 7.3, it doesn't matter), and then proceed with the steps above.

I've tested all sorts of combinations to try and get around this 'catch-22', but nothing works... the above was the only way to get my OpenCV back (after it broke when i updated to MacOS Sierra, 10.12).

**Comment by virajago** [comment]
@danieldf Thanks for the approach. I accidentally came across this thread
[](https://github.com/Homebrew/homebrew-science/issues/4104)

The fix seems to be merged into the code base but one needs to install it from the source. This worked for me ```brew install opencv3 --HEAD --with-contrib```

**Comment by virajago** [comment]
@danieldf Thanks for the detailed approach. 

I accidentally came across this thread which fixes my issue
[](https://github.com/Homebrew/homebrew-science/issues/4104)

The fix seems to be merged into the source but need to install directly from source using this command
`brew install opencv3 --HEAD --with-contrib`

**Comment by fingerco** [comment]
@danieldf thanks for your help.

I think I found the offending line of code that forces us to compile OpenCV manually:

https://github.com/Homebrew/brew/blob/564fa8867dd865c686d243ba48c202e0cb6a35fe/Library/Homebrew/extend/os/mac/diagnostic.rb#L17

That seems to be trying to check if we're on a prerelease version (other methods in the codebase use .prerelease? method for this block of code).

With xCode 7.3 installed, I compiled Cuda like you mentioned. Then, instead of compiling OpenCV by hand, I did the following commands:

```
$ export CI="not_nil"
$ brew install --force opencv3 --HEAD --with-contrib --with-cuda --with-ffmpeg --with-tbb --with-qt5
```

And it allowed me to use brew install OpenCV without checking my Command Line Tools version first.

Whether or not this has unintended side-effects, I will find out in a few hours and update this post accordingly.
